Firestore/core/test/unit/api/load_bundle_task_test.cc

```diff
@@ -0,0 +1,415 @@
+                                   [&](LoadBundleTaskProgress p) {
+                                     EXPECT_EQ(p, error_progress);
+                                     queue.push(1);
+                                   });
+
+  task.SetError();
+
+  EXPECT_EQ(1, queue.pop());
+}
+
+TEST(LoadBundleTaskTest, ProgressObserveTriggers) {
+  LoadBundleTask task(CreateUserQueue());
+  BlockingQueue<int> queue;
+
+  auto handle1 = task.ObserveState(LoadBundleTaskState::InProgress,
+                                   [&](LoadBundleTaskProgress p) {
+                                     EXPECT_EQ(p, Progress(1, 5));
+                                     queue.push(1);
+                                   });
+  auto handle2 = task.ObserveState(LoadBundleTaskState::InProgress,
+                                   [&](LoadBundleTaskProgress p) {
+                                     EXPECT_EQ(p, Progress(1, 5));
+                                     queue.push(2);
+                                   });
+  auto handle3 = task.ObserveState(LoadBundleTaskState::InProgress,
+                                   [&](LoadBundleTaskProgress p) {
+                                     EXPECT_EQ(p, Progress(1, 5));
+                                     queue.push(3);
+                                   });
+
+  task.UpdateProgress(Progress(1, 5));
+
+  EXPECT_EQ(1, queue.pop());
+  EXPECT_EQ(2, queue.pop());
+  EXPECT_EQ(3, queue.pop());
+}
+
+TEST(LoadBundleTaskTest, RemovesAllObservers) {
+  LoadBundleTask task(CreateUserQueue());
+
+  auto handle1 = task.ObserveState(LoadBundleTaskState::InProgress,
+                                   [&](LoadBundleTaskProgress p) {
+                                     (void)p;
+                                     FAIL() << "Removed observer is called.";
+                                   });
+  auto handle2 = task.ObserveState(LoadBundleTaskState::InProgress,
+                                   [&](LoadBundleTaskProgress p) {
+                                     (void)p;
+                                     FAIL() << "Removed observer is called.";
+                                   });
+  auto handle3 = task.ObserveState(LoadBundleTaskState::Error,
+                                   [&](LoadBundleTaskProgress p) {
+                                     (void)p;
+                                     FAIL() << "Removed observer is called.";
+                                   });
+  task.RemoveAllObservers();
+
+  task.UpdateProgress(Progress(1, 5));
+  task.SetError();
+
+  BlockingQueue<int> queue;
+  auto handle4 = task.ObserveState(LoadBundleTaskState::Success,
+                                   [&](LoadBundleTaskProgress p) {
+                                     EXPECT_EQ(p, SuccessProgress());
+                                     queue.push(1);
+                                   });
+  task.SetSuccess(SuccessProgress());
+
+  EXPECT_EQ(1, queue.pop());
+}
+
+TEST(LoadBundleTaskTest, ProgressesFireInOrder) {
```

Comment:
Can you document that the observers are guaranteed to be notified in order in the comments of `LoadBundleTask`?

Reply:
They are not complete ordered, only ordered by state.

We need to determine the semantics for `observe` first.

Reply:
Hmm, but doesn't this test rely on the order being `1, 2, 3`? Yeah, let's discuss the semantics on Monday.

Reply:
They are totally ordered now.